### ISSUE TITLE
Integration of Deep Survival Machines

### ISSUE THREAD

**Issue by chiragnagpal** [havakv/pycox#43]
Hi, I am the first author on the paper, '[Deep Survival Machines: Fully Parametric Survival Regression and Representation Learning for Censored Data with Competing Risks
](https://arxiv.org/abs/2003.01176) and think it might be a good addition to pycox. My current code for DSM is in torch and here.. https://github.com/chiragnagpal/DeepSurvivalMachines 

I'd look forward to beginning discussions regarding integration in case that's of interest...

Cheers !

**Comment by CamDavidsonPilon** [comment]
I have no involvement in PyCox (just a passive observer ;), but I looked over that paper. Cool stuff @chiragnagpal! I really like the graphical abstract in Fig 1.

**Comment by chiragnagpal** [comment]
@CamDavidsonPilon Thanks for giving it a read...  you're right, what we really mean is constant proportional hazards...

**Comment by havakv** [comment]
Hi, the Deep Survival Machines are a very interesting approach and I think they would be a very nice addition to pycox. I'm on vacation for the next couple of weeks, so I probably want be able to look into it before I'm back, but you are more than welcome to start on a pull request.

I haven't fully comprehended the details of the DSM, but a reasonable place to start would be to put the loss functions in [loss.py](https://github.com/havakv/pycox/blob/master/pycox/models/loss.py) and create a [model class](https://github.com/havakv/pycox/tree/master/pycox/models) that contains the necessary logic for training and prediction. Mostly, methods only need a defined `loss` and a `predict_surv` method (e.g., [LogisticHazard](https://github.com/havakv/pycox/blob/master/pycox/models/logistic_hazard.py#L11)) but the DSM probably need something extra for the pre-training? As far as I can understand, you shouldn't need to worry about data loaders, but you should probably include an example network to show how the method works.

I understand that they way pycox is build on torchtuples can make it strange to build a model (probably some things there I should have changed...) so I can only encourage you to try to look at some of the methods and see if you can find a reasonable way to rewrite DSM in this manner. If something is unclear, you shouldn't hesitate to ask me, and I'll try to assist you! Does this sound reasonable?

**Comment by havakv** [comment]
@chiragnagpal have you started on this? Do you need any help with this?

**Comment by chiragnagpal** [comment]
Hi we released a reasonably well documented package with the deep survival machines model here: https://autonlab.github.io/DeepSurvivalMachines . At this stage I do not have enough bandwidth to reepackage it to. be compatible with pycox, folks interested in experimenting with dsm can use our package.

**Comment by havakv** [comment]
Great work! It looks good. I've also realised that I might not have much time in the foreseeable future to work on this. But if I can find the time, I'll ask for your opinion on the code.
